crypto/bio/bio_lib.c

```diff
@@ -578,6 +578,8 @@ long BIO_ctrl(BIO *b, int cmd, long larg, void *parg)
             return ret;
     }
 
+    if (cmd == BIO_C_SET_SSL)
+        BIO_set_shutdown(b, larg);
```

Comment:
Not sure I understand why this is needed, as this is in fact already done in bio_ssl.c:

https://github.com/openssl/openssl/blob/398ae8231650c4bd8ddff0e5efd38233c23b1ca0/ssl/bio_ssl.c#L282-L289

Also, placing a "fix" that belongs there over here is hackery that we should refrain from, as it breaks modularity and becomes harder to maintain.

Reply:
Ah, I did not notice that there was already a call to `BIO_set_shutdown()` for the specific case of SSL BIOs.
So I suppose it would be better to remove the (now redundant) call here,
because `BIO_set_shutdown()` should be called also for other types of BIOs, shouldn't it?

Reply:
So how to proceed here? Remove the general call to `BIO_set_shutdown(b, larg)` for all types of BIOs,
such that only the SSL-specific one remains?

The documentation says:
```
       BIO_set_ssl() sets the internal SSL pointer of BIO b to ssl using the close flag c.
```
which does not explicity say that this call has effect only if `b` is an SSL BIO.
On the other hand, this very call is about setting SSL data, so how about clarifying that this call is meant for SSL BIOs only?

Reply:
IMO the call is meant for SSL BIOs only (those that handle BIO_C_SET_SSL) and should fail otherwise. This is just a documentation omission.

Reply:
Fine, so I've removed again the potential generic call to `BIO_set_shutdown()` from `BIO_ctrl()`,
and in `BIO_f_ssl.pod` made clear where an SSL BIO is expected as an argument, for `BIO_set_ssl() ()` and other functions described there where this was not explicit.